### DISCUSSION TITLE
How should account-transfer transactions work?

### DISCUSSION THREAD

**Discussion by nazrinharris** [nazrinharris/pecunia#64] [Discussion]
Specifically, how would I implement such a feature? How would I store it in the database and ensure its in sync with everything? There's a couple of ways I have thought of:

## 1. Two linked `debit`/`credit` transactions on the respective accounts
This is (according to GPT) is the standard way of doing things. Whenever a transfer is created, a `debit` transaction is created on the "from" account, and a `credit` transaction is created on the "to" account. 

This will require an update on the `Transaction` entity. As explained in #63, there would be a new `TransferDetail` class, of which will hold the `accountId` and `linkedTransactionId`. I'll only be keeping the ids, and if the need for the actual contents arise, I'll retrieve it from the database.

### Pros
1. A relatively simple implementation. It doesn't require a massive overhaul of the current transaction system. 
2. Allows for easy and efficient query-calculations, precisely because there's little change to the current system. I can simply query and sum the transactions, all through SQL.

### Cons
1. The only context a `Transaction` has on whether its a transfer or not would be the `TransferDetail` object. And so if it wants to, say, find out the corresponding transaction and account, it'll need to make a whole retrieval. I doubt that it'll affect performance, but just keep it in mind.

## 2. Create a new `Transfer` table
Basically, this `Transfer` table will be holding all the account-transfer related records. It'll probably hold something like `transferId`, `fromTransactionId`, `fromAccountId`, `toTransactionId`, `toAccountId`. I'm not a big fan of this, at least not yet.

### Pros
1. Encapsulation of transfer logic. It ensures that all transfer records are separate.
2. ... (?)

### Cons
1. Added complexity. Since there's technically two sources of transactions, joins will need to be made if I, for example, want to calculate the balance of an account.
2. ... (?)

**Comment by nazrinharris** [comment]
Used the `TransferDetail` approach. See #63
